File: Java/labs/lab04_Builder/task_3_1/src/CarDirector.java
public class CarDirector {

    private CarBuilder carBuilder;

    public void setCarBuilder(CarBuilder carBuilder) {
        this.carBuilder = carBuilder;
    }

    public Car makeDefaultCar(){
        if (carBuilder == null){
            throw new IllegalArgumentException();
        }
        return carBuilder
                .reset()
                .build();

    }

    public Car makeSportCar(){
        return carBuilder
                .reset()
                .setCarColor(Car.CarColor.Red)
                .setEngine(new Engine(450, 605, Engine.Fuel.Diesel, 3.5f))
                .setWheel(new Wheel(Wheel.Material.Alloy, 21))
                .build();
    }

    public Car makeSUV() {
        return carBuilder
                .reset()
                .setType(Car.Type.SUV)
                .setEngine(new Engine(198, 420, Engine.Fuel.Diesel, 2.2f))
                .setWheel(new Wheel(Wheel.Material.Forged, 19))
                .setCarColor(Car.CarColor.Black)
                .build();
    }
}

File: Cpp/lab05_Prototype/task_2_2/src/car_director.cpp
#include "car_director.hpp"


CarDirector::CarDirector() {}

void CarDirector::setCarBuilder(CarBuilder* carBuilder)
{
  this->carBuilder = carBuilder;
}

Car* CarDirector::makeDefaultCar()
{
  return this->carBuilder
             ->reset()
             ->build();
}

Car* CarDirector::makeSportCar()
{
  return this->carBuilder
             ->reset()
             ->setCarColor(Car::CarColor::Red)
             ->setEngine(new Engine(450, 605, Engine::Fuel::Diesel, 3.5))
             ->setWheel(new Wheel(Wheel::Material::Alloy, 21))
             ->build();
}
Car* CarDirector::makeSUV()
{
  return this->carBuilder
             ->reset()
             ->setType(Car::CarType::SUV)
             ->setEngine(new Engine(198, 420, Engine::Fuel::Diesel, 2.2))
             ->setWheel(new Wheel(Wheel::Material::Forged, 19))
             ->setCarColor(Car::CarColor::Black)
             ->build();
}
